Replaced the encryption with age

raphaelahrens opened the pull request:
The old encryption protocol has some flaws.

In this commit the encryption is replaced by the age library.

raphaelahrens commented:
BTW this is about #35

What is currently missing is how to deal with old encrypted files.
Maybe an alternative tool to decrypt the old files?
maybe it would also be a good idea to have some form of version header, so that this issue does not come up with future changes.

Further could this change also fix #63.
I also found an issue if you run 

```shell
tuckr encrypt g1 ~/first.txt
tuckr encrypt g1 ~/second.txt
```
It is possible to use different passwords for the same file.

This later is an issue when running 

```shell
tuckr decrypt g1
```
Since there you can only give one password and this just fails as soon as the password was wrong.
I have not tested this with the old code.

RaphGL commented:
> What is currently missing is how to deal with old encrypted files.
Maybe an alternative tool to decrypt the old files?
maybe it would also be a good idea to have some form of version header, so that this issue does not come up with future changes.

Yeah, having a header and a tool to port over would be great. We could also provide a single release warning when people attempt to use secrets asking them to port over and maybe provide a link to the tool.

> Further could this change also fix https://github.com/RaphGL/Tuckr/issues/63.

If you want to fix it you'll have to change how `SecretsHandler::try_new` works, right now it asks for the password right away so the user will have to provide a password before it asks for confirmation to override secrets. You could detect all changes.

> It is possible to use different passwords for the same group (g1).

Yeah, secrets is not as developed as the other parts. Symlinks have been consuming most of my time on the project. If you feel like trying to tackle it feel free to do so, otherwise just create an issue and me or someone else will handle it later.

RaphGL commented on Cargo.toml:
```diff
@@ -20,3 +20,4 @@ sha2 = "0.10"
 sys-locale = "0.3.2"
 tabled = {version = "0.10", features = ["color"]}
 zeroize = "1.5"
+age = "0.11"
```
You forgot to remove chacha20poly1305 and sha2 as dependencies

RaphGL commented:
by the way, just as a heads up, don't worry too much about internationalization for now.
I'll go over every output string in the program and make sure it's translated properly before I ship the next tagged release

RaphGL commented:
By the way I haven't attempted to merge your PR because you have it marked as a draft. But there doesn't seem to be any work going on in this. Should I merge?

RaphGL commented:
I'm fine with it either way tbh. But if you want me to do it it's gonna be a while before it's all implemented, since I wanna improve the symlink handling first and I'm working on the [filetree-refactor](https://github.com/RaphGL/Tuckr/tree/filetree-refactor) so I can try and fix the longest standing issues in the project.

raphaelahrens committed:
Replaced the encryption with age

The old encryption protocol has some flaws.

In this commit the encryption is replaced by the age library.

RaphGL commented:
> sorry for taking so long, my priorities for this PR are a abit low

Encryption is also a low priority for me right now, as the core functionality is deploying dotfiles and that still needs some ironing out here and there, so take your time

> This config file will be encrypted with the passphrase.
This has the advantages that we can test the encryption passphrase against all other calls of tuckr encrypt g1 <file> and not have the issue that one file is encrypted with another passphrase.

what happens if the user deletes this config file? Maybe using any one encrypted file in secrets would yield the same result and still enforce the same password per group and if a file is deleted as long as another is there the enforcement of same password per group can still be done.

> Further can the name of the file be hidden by using $path_to_file/base64(HMAC(group_secret, reatl_target_path)).secret and the group secret can be stored in the group file.

Yeah this makes sense, but HMAC is only a hash so how would we know what the name of the file will be once decrypted? When you do `tuckr decrypt x` it will need to be able to compute the destination path, with this method I think the file would have to exist prior to being replaced by the decrypted one so tuckr could compare hashes.

> So the question would be where to store the group files?
What format would you prefer for the group files?

Not sure if I understood correctly. If you're asking what location in the dotfiles, they're still supposed to be in `Secrets` and separated by group directories. 
I don't care too much about what format those groups use or what file structure it follows. The user would never be working directly with it either way.

raphaelahrens commented:
About the path of the group file.
The question was where would I store the group file of `g1`, not the config files in the group.

RaphGL commented:
I'm guessing `v1` is the versioning of the encryption tuckr will use for this group so that we can retain backwards compatibility in the future? 

For the format itself, since the config will be inside a group directory I think we can get rid of `group_name`, `group_secret` can just be named `secret` and we can infer the group name, this also allows people to just rename the directory and is consistent with how the program works everywhere else. 

For the config being deleted, if we could detect the encryption used by each file we could just attempt to read the config but if it's absent we could attempt to restore it and check that all the files still use the same algo? Not sure if that's a good strategy or not.

RaphGL committed:
Merge branch 'master' into rage_port

ThomasAdam commented:
> But once I get around to it, I'll choose what to do. I don't even use secrets myself. So unless there's interest in this, this might end up getting the axe in the future.

Sure.  I can understand this.  I can also appreciate how this looked bed.
